File: string/464-groupAnagram.cpp
#include<string>
#include<vector>
#include<iostream>
#include<map>
#include<algorithm>  //for std::sort
using STR=std::string;
using VEC=std::vector<STR>;
using MAP=std::map<STR,VEC>;
using MATRIX=std::vector<std::vector<STR>>;
// void _counter()

MATRIX groupString(const VEC& s)
{
    MAP data{};
    MATRIX res{};
    for(auto _s:s)
    {
        STR tmp=_s;
        std::sort(tmp.begin(),tmp.end());
        data[tmp].push_back(_s);
    }
    
    for(auto &[k,v]:data)
    {
        res.push_back(v);
    }
    return res;
}

int main()
{
    VEC strs{"eat","tea","tan","ate","nat","bat"};
    MATRIX ans=groupString(strs);
    for(auto &ss:ans)
    {
        std::cout<<'\n';
        for(auto &s:ss)
            std::cout<<s<<',';
    }
}

File: string/464-groupAnagram.py
#https://stackoverflow.com/questions/6754102/typeerror-unhashable-type
from collections import defaultdict
'''
an array of strings strs, group the anagrams together. You can return the answer in any order. An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

Example 1:
Input: strs = ["eat","tea","tan","ate","nat","bat"]
Output: [["bat"],["nat","tan"],["ate","eat","tea"]]

Example 2:
Input: strs = [""]
Output: [[""]]

Example 3:
Input: strs = ["a"]
Output: [["a"]]
'''
def groupAnagrams(strs):
    """
    :type strs: List[str]
    :rtype: List[List[str]]
    """
    #defaultdict
    data=defaultdict(list)
    for w in strs:
        data[tuple(sorted(w))].append(w)
    return data.values()

def groupAnagrams2(strs):
    data=defaultdict(list)
    for w in strs:
        d=[0]*26
        for c in w:
            d[ord(c)-97]+=1
        data[tuple(d)].append(w)
    return data.values()

strs = ["eat","tea","tan","ate","nat","bat"]
print(list(groupAnagrams(strs)))
print(list(groupAnagrams2(strs)))
